Microsoft.Azure.Cosmos.Encryption/src/MdeSupport/EncryptionKeyStoreProviderImpl.cs

```diff
@@ -0,0 +1,65 @@
+﻿//------------------------------------------------------------
+// Copyright (c) Microsoft Corporation.  All rights reserved.
+//------------------------------------------------------------
+
+namespace Microsoft.Azure.Cosmos.Encryption
+{
+    using System;
+    using Microsoft.Data.Encryption.Cryptography;
+
+    /// <summary>
+    /// The purpose/intention to introduce this class is to utilize the cache provide by the <see cref="EncryptionKeyStoreProvider"/> abstract class. This class basically
+    /// redirects all the corresponding calls to <see cref="CosmosEncryptionKeyStoreProvider"/> 's overridden methods and thus allowing us
+    /// to utilize the virtual method <see cref="EncryptionKeyStoreProvider.GetOrCreateDataEncryptionKey"/> to access the cache.
+    ///
+    /// Note: Since <see cref="EncryptionKeyStoreProvider.Sign"/> and <see cref="EncryptionKeyStoreProvider.Verify"/> methods are not exposed, <see cref="EncryptionKeyStoreProvider.GetOrCreateSignatureVerificationResult"/> is not supported either.
+    /// </summary>
+    internal sealed class EncryptionKeyStoreProviderImpl : EncryptionKeyStoreProvider
+    {
+        private readonly CosmosEncryptionKeyStoreProvider cosmosEncryptionKeyStoreProvider;
+
+        public EncryptionKeyStoreProviderImpl(CosmosEncryptionKeyStoreProvider cosmosEncryptionKeyStoreProvider)
+        {
+            this.cosmosEncryptionKeyStoreProvider = cosmosEncryptionKeyStoreProvider;
+        }
+
+        public override string ProviderName => this.cosmosEncryptionKeyStoreProvider.ProviderName;
+
+        public override byte[] UnwrapKey(string encryptionKeyId, KeyEncryptionKeyAlgorithm algorithm, byte[] encryptedKey)
+        {
+            return this.cosmosEncryptionKeyStoreProvider.UnwrapKeyAsync(encryptionKeyId, algorithm.ToString(), encryptedKey)
+                .ConfigureAwait(false)
+                .GetAwaiter()
+                .GetResult();
+        }
+
+        public override byte[] WrapKey(string encryptionKeyId, KeyEncryptionKeyAlgorithm algorithm, byte[] key)
+        {
+            return this.cosmosEncryptionKeyStoreProvider.WrapKeyAsync(encryptionKeyId, algorithm.ToString(), key)
+                .ConfigureAwait(false)
+                .GetAwaiter()
+                .GetResult();
+        }
+
+        public new byte[] GetOrCreateDataEncryptionKey(string encryptedDataEncryptionKey, Func<byte[]> createItem)
```

Comment:
To avoid confusion and possible bugs in the future can this be renamed? The danger of using new is when it gets casted down to EncryptionKeyStoreProvider it will just call the base implementation. If any logic gets added to this in the future its possible it will not be called. This has caused bugs in the past.

```suggestion
        public new byte[] GetOrCreateDataEncryptionKeyHelper(string encryptedDataEncryptionKey, Func<byte[]> createItem)
```